error.go

```diff
@@ -45,27 +67,90 @@ func (e Errs) Error() string {
 	return fmt.Sprintf("%d error(s) occurred.\n\t%s", len(e), strings.Join(details, "\n\t"))
 }
 
-// Error returns a string containing the region, service as well as the original API error message.
-func (e *callErr) Error() string {
-	return fmt.Sprintf("region: %s, service: %s, Error message: %q",
-		e.region,
-		e.service,
-		e.err.Error())
-}
+// ErrorIn inspects err to find if there is an error in the region and returns it
+// or them (in case of multiple), otherwise it returns nil.
+// err must be of the type Error or Errors in order to be able to find if there
+// are errors in the region; if err is from other type, the function always
+// returns nil.
+// The returned error is a value of the type Error when only one error is found
+// in the region, or a value of the type Errors when multiple errors are found.
+func ErrorIn(region string, err error) error {
+	switch e := err.(type) {
+	case Error:
+		if e.Region() == region {
+			return err
+		}
 
-// Returns the region of the error
-func (e *callErr) Region() string {
-	return e.region
-}
+		return nil
+	case *Error:
+		if e == nil {
+			return nil
+		}
 
-// Returns the service name of the error
-func (e *callErr) Service() string {
-	return e.service
+		if e.Region() == region {
+			return err
+		}
+
+		return nil
+	case Errors:
+		var errs Errors
+
+		for _, ei := range e {
+			if ei.Region() == region {
+				errs = append(errs, ei)
+			}
+		}
+
+		if len(errs) > 0 {
+			return errs
+		}
+
+		return nil
+	default:
+		return nil
+	}
 }
 
-type callErr struct {
-	Err
-	err     error
-	region  string
-	service string
+// ErrorFrom inspects err to find if there is an error in the service and returns
+// it or them (in case of multiple), otherwise it returns nil.
+// err must be of the type Error or Errors in order to be able to find if there
+// are errors in the region; if err is from other type, the function always
+// returns nil.
+// The returned error is a value of the type Error when only one error is found
+// in the region, or a value of the type Errors when multiple errors are found.
+func ErrorFrom(service string, err error) error {
+	switch e := err.(type) {
+	case Error:
+		if e.Service() == service {
+			return err
+		}
+
+		return nil
+	case *Error:
+		if e == nil {
+			return nil
+		}
+
+		if e.Service() == service {
+			return err
+		}
+
+		return nil
+	case Errors:
+		var errs Errors
+
+		for _, ei := range e {
+			if ei.Service() == service {
+				errs = append(errs, ei)
+			}
+		}
```

Comment:
Same comment about recursion.

Reply:
The logic is tiny, it's basically a condition and doing a recursion doesn't convince me

```
      for _, ei := range e 
		if ef := ErrorFrom(service, ei); ef != nil {
         	errs = append(errs, ef.(Error))
        }
      }
```

There is a conditional anyway and a type assertion; considering the replicated logic, I see that's more readable the loop version.

Without considering that, historically, Go behaves better in loops than recursion calls and that the each recursive call has to execute the `switch type` again, however these 2 things are not an argument without providing benchmarks.